Auto client

dkosteck opened the pull request:
Allowing for automatic runs from the Python client, including specific configs. This client should be standalone, and can be used outside of this repo if desired (we can discuss how best we want to accomplish this based on our discussion of further testing).

akiselev1 commented on rfc2544/README.md:
```diff
@@ -174,3 +174,35 @@ Returns a dict, with results if available, empty otherwise.
+
+This mode will run with all default values if no other arguments are set. However, please note that the `timeout` is specific to auto mode and should be adjusted if timeouts are observed. This mode also allows for bypassing the majority of the CLI arguments when `config` is set. This will look for a JSON formatted config file which contains the arguments one wishes to override. The example templa ...
+
+Upon successful run auto mode should return a single line to stdout: the result JSON.
```
Can we add  "auto" client command line sample and json output to the doc?

dkosteck replied:
Yup, will do

jianzzha commented on rfc2544/client.py:
```diff
@@ -2,50 +2,141 @@
+        "runtime_tolerance": args.runtime_tolerance,
+        "negative_packet_loss": args.negative_packet_loss,
+        "rate_tolerance_failure": args.rate_tolerance_failure,
+        "binary_search_extra_args": [],
```
I have been using these options and they are missing here, they are important if we want to have good performance reading on the first trial: --send-teaching-warmup \
            --teaching-warmup-packet-type=generic \
            --teaching-warmup-packet-rate=10000

dkosteck replied:
Okay, I will add them

dkosteck commented:
Updated to address comments and improvements: 
- Takes in the server port as argument
- Added missing trafficgen args to client
- Added client examples to README

jianzzha commented:
@akiselev1 Please test out this client with the 710 connection against the dell switch and see if you can reach the line rate

akiselev1 commented:
> @akiselev1 Please test out this client with the 710 connection against the dell switch and see if you can reach the line rate

Results look the same with or without client: 
```
[root@53X8R53 rfc2544]# python3 client.py auto --server-addr 127.0.0.1 --server-port 8080
{'0': {'rx_l1_bps': 21290049228.870625, 'rx_l2_bps': 15207178020.621874, 'rx_latency_average': 6.0, 'rx_latency_maximum': 22.0, 'rx_latency_minimum': 5.0, 'rx_pps': 31681620.87629557, 'tx_l1_bps': 21290049228.870625, 'tx_l2_bps': 15207178020.621874, 'tx_pps': 31681620.87629557}, '1': {'rx_l1_bps': 21290049228.870625, 'rx_l2_bps': 15207178020.621874, 'rx_latency_average': 7.0, 'rx_latency_maximum': 22.0, 'rx_latency_minimum': 6.0, 'rx_pps': 31681620.87629557, 'tx_l1_bps': 21290049228.870625, 'tx_l2_bps': 15207178020.621874, 'tx_pps': 31681620.87629557}}
[root@53X8R53 rfc2544]#
```
The only thought I had looking at the client output: can we include rx_lost_packets into json? that would make it easier to interpret the result.

dkosteck commented:
> > @akiselev1 Please test out this client with the 710 connection against the dell switch and see if you can reach the line rate
> 
> Results look the same with or without client:
> 
> ```
> [root@53X8R53 rfc2544]# python3 client.py auto --server-addr 127.0.0.1 --server-port 8080
> {'0': {'rx_l1_bps': 21290049228.870625, 'rx_l2_bps': 15207178020.621874, 'rx_latency_average': 6.0, 'rx_latency_maximum': 22.0, 'rx_latency_minimum': 5.0, 'rx_pps': 31681620.87629557, 'tx_l1_bps': 21290049228.870625, 'tx_l2_bps': 15207178020.621874, 'tx_pps': 31681620.87629557}, '1': {'rx_l1_bps': 21290049228.870625, 'rx_l2_bps': 15207178020.621874, 'rx_latency_average': 7.0, 'rx_latency_maximum': 22.0, 'rx_latency_minimum': 6.0, 'rx_pps': 31681620.87629557, 'tx_l1_bps': 21290049228.870625, 'tx_l2_bps': 15207178020.621874, 'tx_pps': 31681620.87629557}}
> [root@53X8R53 rfc2544]#
> ```
> 
> The only thought I had looking at the client output: can we include rx_lost_packets into json? that would make it easier to interpret the result.

I'll take a look at why it's not being pulled from the report, sure

dkosteck commented:
@akiselev1 I just pushed an update with the full report through the REST API, can you please run it again and let me know if this now works? (it should include all fields from the report)

akiselev1 commented:
> @akiselev1 I just pushed an update with the full report through the REST API, can you please run it again and let me know if this now works? (it should include all fields from the report)

Looks good now: 
```
[root@53X8R53 rfc2544]# python3 client.py auto --server-addr 127.0.0.1 --server-port 8080
{'0': {'rx_active': True, 'rx_l1_bps': 20899358218.506004, 'rx_l2_bps': 14928113013.218575, 'rx_latency_average': 7.0, 'rx_latency_l1_bps': 454639.9056256746, 'rx_latency_l2_bps': 324742.7897326247, 'rx_latency_lost_pps': 0.0, 'rx_latency_maximum': 19.0, 'rx_latency_minimum': 6.0, 'rx_latency_packets': 30001, 'rx_latency_pps': 676.5474786096348, 'rx_lost_packets': 0, 'rx_lost_packets_pct': 0.0, 'rx_lost_pps': 0.0, 'rx_packets': 1379117051, 'rx_pps': 31100235.444205362, 'tx_active': True, 'tx_l1_bps': 20899358218.506004, 'tx_l2_bps': 14928113013.218575, 'tx_latency_l1_bps': 454639.9056256746, 'tx_latency_l2_bps': 324742.7897326247, 'tx_latency_packets': 30001, 'tx_latency_pps': 676.5474786096348, 'tx_packets': 1379117051, 'tx_pps': 31100235.444205362, 'tx_pps_target': 45970569.89397322, 'tx_tolerance_max': 68955854.84095982, 'tx_tolerance_min': 22985284.94698661}, '1': {'rx_active': True, 'rx_l1_bps': 20899358218.506004, 'rx_l2_bps': 14928113013.218575, 'rx_latency_average': 7.0, 'rx_latency_l1_bps': 454639.9056256746, 'rx_latency_l2_bps': 324742.7897326247, 'rx_latency_lost_pps': 0.0, 'rx_latency_maximum': 20.0, 'rx_latency_minimum': 5.0, 'rx_latency_packets': 30001, 'rx_latency_pps': 676.5474786096348, 'rx_lost_packets': 0, 'rx_lost_packets_pct': 0.0, 'rx_lost_pps': 0.0, 'rx_packets': 1379117051, 'rx_pps': 31100235.444205362, 'tx_active': True, 'tx_l1_bps': 20899358218.506004, 'tx_l2_bps': 14928113013.218575, 'tx_latency_l1_bps': 454639.9056256746, 'tx_latency_l2_bps': 324742.7897326247, 'tx_latency_packets': 30001, 'tx_latency_pps': 676.5474786096348, 'tx_packets': 1379117051, 'tx_pps': 31100235.444205362, 'tx_pps_target': 45970569.89397322, 'tx_tolerance_max': 68955854.84095982, 'tx_tolerance_min': 22985284.94698661}}
[root@53X8R53 rfc2544]# 
```
